src/sage/schemes/elliptic_curves/addition_formulas_ring.py

```diff
@@ -0,0 +1,64 @@
+
+def add(E, P, Q):
+    r"""
+    Addition formulas for elliptic curves over general rings
+    with trivial Picard group.
+
+    REFERENCES:
+
+    These formulas were derived by Bosma and Lenstra [BL1995]_,
+    with corrections by Best [Best2021]_ (Appendix A).
+
+    EXAMPLES::
+
+        sage: from sage.schemes.elliptic_curves.addition_formulas_ring import add
+        sage: M = Zmod(13*17*19)
+        sage: R.<U,V> = M[]
+        sage: S.<u,v> = R.quotient(U*V - 17)
+        sage: E = EllipticCurve(S, [1,2,3,4,5])
+        sage: P = E(817, 13, 19)
+        sage: Q = E(425, 123, 17)
+        sage: PQ1, PQ2 = add(E, P, Q)
+        sage: PQ1
+        (1188, 1674, 540)
+        sage: PQ2
+        (582, 2347, 1028)
+        sage: E(PQ1) == E(PQ2)
+        True
+    """
+    a1, a2, a3, a4, a6 = E.a_invariants()
+    b2, b4, b6, b8 = E.b_invariants()
+
+    assert P in E
+    assert Q in E
+    X1, Y1, Z1 = P
+    X2, Y2, Z2 = Q
+
+    # TODO: I've made a half-hearted attempt at simplifying the formulas
+    # by caching common subexpressions. This could almost certainly be
+    # sped up significantly with some more serious optimization effort.
+
+    XYdif = X1*Y2 - X2*Y1
+    XYsum = X1*Y2 + X2*Y1
+    XZdif = X1*Z2 - X2*Z1
+    XZsum = X1*Z2 + X2*Z1
+    YZdif = Y1*Z2 - Y2*Z1
+    YZsum = Y1*Z2 + Y2*Z1
+
+    a1sq, a2sq, a3sq, a4sq = (a**2 for a in (a1, a2, a3, a4))
+
+    X31 = XYdif*YZsum+XZdif*Y1*Y2+a1*X1*X2*YZdif+a1*XYdif*XZsum-a2*X1*X2*XZdif+a3*XYdif*Z1*Z2+a3*XZdif*YZsum-a4*XZsum*XZdif-3*a6*XZdif*Z1*Z2
+
+    Y31 = -3*X1*X2*XYdif-Y1*Y2*YZdif-2*a1*XZdif*Y1*Y2+(a1sq+3*a2)*X1*X2*YZdif-(a1sq+a2)*XYsum*XZdif+(a1*a2-3*a3)*X1*X2*XZdif-(2*a1*a3+a4)*XYdif*Z1*Z2+a4*XZsum*YZdif+(a1*a4-a2*a3)*XZsum*XZdif+(a3sq+3*a6)*YZdif*Z1*Z2+(3*a1*a6-a3*a4)*XZdif*Z1*Z2
+
+    Z31 = 3*X1*X2*XZdif-YZsum*YZdif+a1*XYdif*Z1*Z2-a1*XZdif*YZsum+a2*XZsum*XZdif-a3*YZdif*Z1*Z2+a4*XZdif*Z1*Z2
+
+    yield (X31, Y31, Z31)
```

Comment:
```suggestion
```

Using `yield` makes this function a generator which leads to some unintuitive behaviour.

For instance,

```python
p_plus_q = add(E, P, Q)
PQ1, PQ2 = p_plus_q
PQ3, PQ4 = p_plus_q  # This will give an error since the generator is exhausted
```

Reply:
This function is not intended for users. They should just write `P + Q` instead.

Reply:
Okay. If this is meant to be internal then perhaps rename to `_add`? Names starting with an underscore are considered internal: https://doc.sagemath.org/html/en/developer/coding_in_python.html#deprecation